### PR TITLE
a clean commit for #383

### PR DESCRIPTION
**Pull request by wkliao**
This PR contains only my changes for #383.

### CODE REVIEW

**Comment by WardF** [comment]
Looks good I will see how quickly I can get this in, we are trying to get 4.6.0 (previously 4.5.1, but with compression plugin support we have bumped to 4.6.0) out ASAP.  I see the checks failed, will investigate in AM.

File: include/ncdispatch.h [hunk 1]
```diff
@@ -9,7 +9,9 @@
 #ifndef _DISPATCH_H
 #define _DISPATCH_H
 
+#if HAVE_CONFIG_H
```
Comment (edhartnett) [inline include/ncdispatch.h]: We always have config.h, right?

File: libdispatch/dvar.c [hunk 2]
```diff
@@ -538,6 +538,79 @@ NC_getshape(int ncid, int varid, int ndims, size_t* shape)
...
+
+      index[0] = 1;
+      if (nc_put_var1_ushort(ncid, varid, index, &ushort_data)) ERR;
+
+      index[0] = 0;
+      if (nc_get_var1_ushort(ncid, varid, index, &ushort_data_in)) ERR;
+
+      if (nc_close(ncid)) ERR;
+\endcode
+*/
```
Comment (edhartnett) [inline libdispatch/dvar.c]: You should put an \author tag with your name for new functions that you introduce. We're trying to do a better job of maintaining attribution of authorship in the code.

**Comment by edhartnett** [comment]
Awesome, thanks. Much better than the other PR! ;-)

@WardF if you are going to include this in 3.6.0 then I think we will need a little time after this merge is complete to look into all the other reported fill value issues and make sure all is working as it should. For example I have some new tests that will need to be modified to check per-variable fill mode instead of per-file fill mode for classic formats.

**Comment by edhartnett** [comment]
@wkliao after this PR is merged, what happens when the user calls nc_set_fill() to try and set fill value for the whole file. Does that still work?

File: include/ncdispatch.h [hunk 3]
```diff
@@ -250,6 +252,7 @@ int (*inq_var_all)(int ncid, int varid, char *name, nc_type *xtypep,
               );
 
 int (*var_par_access)(int, int, int);
+int (*def_var_fill)(int, int, int, const void*);
```
Comment (edhartnett) [inline include/ncdispatch.h]: Was there a reason to move this line?

File: libdap2/ncd2dispatch.c [hunk 4]
```diff
@@ -2434,6 +2434,15 @@ NCD2_var_par_access(int ncid, int p2, int p3)
     return THROW(NC_ENOPAR);
 }
 
+int
```
Comment (edhartnett) [inline libdap2/ncd2dispatch.c]: Was there a reason to move this function in the file? Or was something else done?

File: libdap2/ncd2dispatch.h [hunk 5]
```diff
@@ -158,6 +158,9 @@ NCD2_rename_var(int ncid, int varid, const char *name);
 extern int
 NCD2_var_par_access(int, int, int);
 
+extern int
+NCD2_def_var_fill(int, int, int, const void *);
+
```
Comment (edhartnett) [inline libdap2/ncd2dispatch.h]: Again, why was this moved? Seems like at least three files are changing for no real reason. Or am I missing something?

Comment (edhartnett) [reply libdap2/ncd2dispatch.h]: OK, never mind. I see that you are moving things outside the #ifdef USE_NETCDF4. Now I get it. ;-)

**Comment by edhartnett** [comment]
OK, to answer my own question, after this PR, when nc_set_fill() is called it will correctly set the fill value of all variables.

File: include/ncdispatch.h (hunk 1 above)
Comment (wkliao) [reply include/ncdispatch.h]: This is a common practice when using autotools.

**Comment by WardF** [comment]
I am reviewing pull requests now and trying to prioritize them, I'd like to include this but we shall see, I also need to get 4.6.0 out to revert some behavior that was introduced in 4.5.0.  So.. maybe. :)

**Comment by WardF** [comment]
Thanks!
